posts: New release! (#302)
gtk-rs/gtk-rs.github.io: _posts/2023-08-28-new-release.md

---
layout: post
author: gtk-rs developers
title: New Release
categories: [front, crates]
date: 2023-08-28 20:00:00 +0000
---

* Write intro here *

TODO:

> - Mention MRSV

Reviewer: It time again for a new gtk-rs release!

If you have been watching our crates on crates.io closely or if you have attended GUADEC you may already have noticed that we have released new versions of gtk-rs crates.

The new versions contain less breaking changes than some of our previous releases, so updating will be easier. The release however does include some nice improvements that can simplify you application as well as support the latest API additions of the underlying libraries.

As anticipated in the previous release, this will likely be the last release of the gtk3-rs bindings and we recommend to everyone to upgrade to GTK4.
If you require gtk3-rs to keep up with new versions of gtk-rs-core please consider volunteering to help with maintenance of the project.

# Increase of the minimum supported Rust version (MSRV)

With this release, Rust 1.70 is the minimum version required to compile and use the bindings.

- Mention the minimum supported versions of each crates / newly supported versions

> - Mention the ValueDelegate macro ?

Reviewer 2: ### New `glib::ValueDelegate` macro

Let's say you want to create a wrapper around some object, but you want to retain `ToValue`, `FromValue`, and `HasParamSpec` implementations, which are especially necessary when you want to use an object as a property. This is where the `glib::ValueDelegate` macro comes in; instead of having to manually implement the following:

```rust
pub struct Uid(String);

impl glib::types::StaticType for Uid {
    fn static_type() -> glib::types::Type {
        <String as glib::types::StaticType>::static_type()
    }
}

impl glib::value::ToValue for Uid {
    fn to_value(&self) -> glib::value::Value {
        let this = self;
        glib::value::ToValue::to_value(&this.0)
    }

    fn value_type(&self) -> glib::types::Type {
        let this = self;
        glib::value::ToValue::value_type(&this.0)
    }
}

impl From<Uid> for glib::value::Value {
    fn from(this: Uid) -> Self {
        glib::value::Value::from(this.0)
    }
}

unsafe impl<'a> glib::value::FromValue<'a> for Uid {
    type Checker = <String as glib::value::FromValue<'a>>::Checker;
    unsafe fn from_value(value: &'a glib::value::Value) -> Self {
        Uid(<String as glib::value::FromValue<'a>>::from_value(value))
    }
}

impl glib::HasParamSpec for Uid {
    type ParamSpec = <String as glib::HasParamSpec>::ParamSpec;
    type SetValue = Self;
    type BuilderFn = <String as glib::HasParamSpec>::BuilderFn;
    fn param_spec_builder() -> Self::BuilderFn {
        <String as glib::HasParamSpec>::param_spec_builder()
    }
}
```

you can simply use the macro as follows

```rust
#[derive(glib::ValueDelegate)]
pub struct Uid(String);
```

For more complex cases, see the [documentation](https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/derive.ValueDelegate.html).

Reviewer: I do not think we should use `object` in the text, since that is ambiguous and seems to refer to `GObject`s. Maybe use `type` or `newtype`

Author: Thank you

- Mention the new docs improvements of subclassing traits / class methods
- Mention the finally working docs on docs.rs for libs depending on gtk-rs

## gtk-rs-core

### New `glib::derived_properties` macro

> Since the addition of `glib::Properties`, you would still have to write this bits manually

Reviewer 3: ```suggestion
With the `glib::Properties` derive macro added in the last release, you still had to write these blanket functions manually
```

```rust
impl ObjectImpl for T {
    fn properties() -> &'static [glib::ParamSpec] {
        Self::derived_properties()
    }
    fn set_property(&self, id: usize, value: &glib::Value, pspec: &glib::ParamSpec) {
        self.derived_set_property(id, value, pspec)
    }
    fn property(&self, id: usize, pspec: &glib::ParamSpec) -> glib::Value {
        self.derived_property(id, pspec)
    }
}
```

> Which you can replace with

Reviewer 3: ```suggestion
Now you can replace that code with
```

```rust
#[glib::derived_properties]
impl ObjectImpl for T {}
```

### Replacement of `glib::Continue` / `glib::Inhibit`

> These two types were a bit difficult to understand how they should be used, as you had to pass a boolean `glib::Continue(true)` /  `glib::Inhibit(false)`. We replaced both with

Reviewer 3: ```suggestion
These two types were a bit difficult to understand, as you had to pass a boolean `glib::Continue(true)` /  `glib::Inhibit(false)`. We replaced them with new enums
```

```rust
glib::Continue(true) -> glib::ControlFlow::Continue
glib::Continue(false) -> glib::ControlFlow::Break
glib::Inhibit(true) -> glib::Propagation::Stop
glib::Inhibit(false) -> glib::Propagation::Proceed
```

Making the meaning of each value clearer to the developer. This should be the most annoying change you need to do during the update to this release.

### Typed constructors

> In the past we had a bunch of constructors taking `glib::Type` as a parameter instead of `T: impl StaticType` / `T: impl IsA<O>`. During this release, we tried to clean few of those constructors.

Reviewer 3: ```suggestion
In the past we had a bunch of constructors taking `glib::Type` as a parameter instead of `T: impl StaticType` / `T: impl IsA<O>`. For this release, we cleaned up several of those constructors.
```

```rust
let model = gio::ListStore::new(SomeObject::static_type()); // before
let model = gio::ListStore::new::<SomeObject>(); // after
let model = gio::ListStore::with_type(SomeObject::static_type()); // is available for specific use cases
```

Changed constructors:
 - `glib::SignalGroup`
 - `glib::FlagsClass`
 - `glib::EnumClass`

> ### Various `glib::Properties` improvements

Reviewer 3: Is this section TODO? There's one improvement already listed at the top.

Author: right, i left a todo there as there were a bunch of small but important enough to note fixes

## gtk4-rs

> The GTK4 bindings didn't see that many user facing changes other then various cleanups and bugfixes. The bindings now support GTK 4.12 with `v4_12` feature flag.

Reviewer 3: ```suggestion
The bindings now support GTK 4.12 with `v4_12` feature flag. Other than that they didn't see many user facing changes. The development mainly focused on various cleanups and bug fixes.
```

### Blueprint support in `gtk::CompositeTemplate`

> You can now use [blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) with both `string` and `file` attributes of `gtk::CompositeTemplate` macro. It requires having the `blueprint-compiler` installed.

Reviewer 3: ```suggestion
You can now use [blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) with both `string` and `file` attributes of the `gtk::CompositeTemplate` macro. This requires having the `blueprint-compiler` binary installed.
```

```rust

> 

Reviewer 3: ```suggestion
```

    #[template(string = "
    template MyWidget : Widget {
        Label label {
            label: 'foobar';
        }

        Label my_label2 {
            label: 'foobaz';
        }
    }
    ")]
    pub struct MyWidget {
        #[template_child]
        pub label: TemplateChild<gtk::Label>,
        #[template_child(id = "my_label2")]
        pub label2: gtk::TemplateChild<gtk::Label>,
    }

    #[glib::object_subclass]
    impl ObjectSubclass for MyWidget {
        const NAME: &'static str = "MyWidget";
        type Type = super::MyWidget;
        type ParentType = gtk::Widget;
        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
        }
        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for MyWidget {
        fn dispose(&self) {
            while let Some(child) = self.obj().first_child() {
                child.unparent();
            }
        }
    }
    impl WidgetImpl for MyWidget {}
}

glib::wrapper! {
    pub struct MyWidget(ObjectSubclass<imp::MyWidget>) @extends gtk::Widget;
}
```

### Stop using deprecated StyleContext APIs

> The `gtk::StyleContext::add_provider_for_display` / `gtk::StyleContext::remove_provider_for_display` were never supposed to be part of the `gtk::StyleContext` type but as separate methods. As the `gtk::StyleContext` was deprecated upstream in GTK 4.10, it became annoying for the developer having to use `#[allow(deprecated)]` even if those two functions were not deprecated at all.

Reviewer 3: ```suggestion
The `gtk::StyleContext::add_provider_for_display` / `gtk::StyleContext::remove_provider_for_display` functions were never supposed to be part of the `gtk::StyleContext` type. Since that type was deprecated upstream in GTK 4.10, the developers now annoyingly had to use `#[allow(deprecated)]`, even though those two functions were not deprecated at all.
```

> We have added `gtk::style_context_add_provider_for_display` / `gtk::style_context_remove_provider_for_display` as a replacement of these.

Reviewer 3: ```suggestion
We have added `gtk::style_context_add_provider_for_display` / `gtk::style_context_remove_provider_for_display` as a replacement for them.
```

> ## Changes

Reviewer 3: ```suggestion
## Detailed Changes
```

> For the interested ones, here is the list of the merged pull requests:

Reviewer 3: ```suggestion
Here is the full list of the merged pull requests:
```

---
layout: post
author: gtk-rs developers
title: New Release
categories: [front, crates]
date: 2023-08-28 20:00:00 +0000
---

TODO:

> - Mention the minimum supported versions of each crates / newly supported versions

Reviewer: Is this just bitflags 2? Or is there anything else? For bitflags shall we link https://github.com/bitflags/bitflags/issues/262 or is there a better migration

If it is just bitflags, I propose:

## Switch to bitflags 2.0

With this release we switched to [version 2](https://github.com/bitflags/bitflags/releases/tag/2.0.0) of the `bitflags` crate. While we regularly update our dependencies to the latest version, this is of particular note since flag types are often use in glib and GTK API.

Reviewer: Oh I see :)

With that said, the bitflags API change is important, so it might be worth to use the paragraph anyway

Author: Included

        <String as glib::types::StaticType>::static_type()
    }
}

impl glib::value::ToValue for Uid {
    fn to_value(&self) -> glib::value::Value {
        glib::value::ToValue::to_value(&self.0)
    }

    fn value_type(&self) -> glib::types::Type {
        glib::value::ToValue::value_type(&self.0)
    }
}

impl From<Uid> for glib::value::Value {
    fn from(uid: Uid) -> Self {
        glib::value::Value::from(uid.0)
    }
}

unsafe impl<'a> glib::value::FromValue<'a> for Uid {
    type Checker = <String as glib::value::FromValue<'a>>::Checker;
    unsafe fn from_value(value: &'a glib::value::Value) -> Self {
        Uid(<String as glib::value::FromValue<'a>>::from_value(value))
    }
}

impl glib::HasParamSpec for Uid {
    type ParamSpec = <String as glib::HasParamSpec>::ParamSpec;
    type SetValue = Self;
    type BuilderFn = <String as glib::HasParamSpec>::BuilderFn;
    fn param_spec_builder() -> Self::BuilderFn {
        <String as glib::HasParamSpec>::param_spec_builder()
    }
}
```

you can simply use the macro as follows:

```rust
#[derive(glib::ValueDelegate)]
pub struct Uid(String);
```

For more complex cases, see the documentation.

> You could also achieve the same by using `glib::Boxed`, but `glib::ValueDelegate` provides a thinner wrapper as it only calls to the inner implementation instead of using [`GBoxed`](https://docs.gtk.org/gobject/boxed.html). Additionally, `glib::ValueDelegate` propagates the `GType`` to the wrapper instead of creating a new type.

Reviewer 2: ```suggestion
You could also achieve the same by using `glib::Boxed`, but `glib::ValueDelegate` provides a thinner wrapper as it only calls to the inner implementation instead of using [`GBoxed`](https://docs.gtk.org/gobject/boxed.html). Additionally, `glib::ValueDelegate` propagates the `GType` to the wrapper instead of creating a new `GType`.
```
